challenge-two/src/agent.ts

```diff
@@ -31,26 +33,57 @@ export const provideHandleTransaction = (
 
     for (const log of swapLogs) {
       try {
+        const poolAddress = log.address.toLowerCase();
         const poolContract = new ethers.Contract(
-          log.address,
+          poolAddress,
           UNISWAP_V3_POOL_ABI,
           provider
         );
 
-        // Check if the log is from a valid Uniswap V3 pool
-        const poolFactory: string = await poolContract.callStatic.factory();
+        // Validate the pool's legitimacy by confirming it was created by the Uniswap V3 factory
+        let token0, token1, fee;
+        try {
+          token0 = await poolContract.token0();
+          token1 = await poolContract.token1();
+          fee = await poolContract.fee();
+        } catch (err) {
+          console.error(
+            `Error fetching token details from pool at ${poolAddress}:`,
+            err
+          );
+          continue; // Skip this pool if fetching details fails
+        }
 
-        if (
-          poolFactory.toLowerCase() !== UNISWAP_V3_FACTORY_ADDRESS.toLowerCase()
-        ) {
-          // This is not a valid Uniswap V3 pool
+        let derivedPoolAddress;
+        try {
+          derivedPoolAddress = await factoryContract.getPool(
+            token0,
+            token1,
+            fee
+          );
+        } catch (err) {
+          console.error(
+            `Error verifying pool creation at ${poolAddress}:`,
+            err
+          );
+          continue; // Skip this pool if verification fails
+        }
+
+        if (derivedPoolAddress.toLowerCase() !== poolAddress.toLowerCase()) {
```

Comment:
Hey Mohamad, yes, the logic up until this point will work, great job!
I will challenge you further though. 😎 

As you can see, you currently have the calls to the pool contract to fetch `token0`, `token1`, and `fee`. _Then_ you have the calls to the Factory contract for `getPool()`.

I challenge you to update the implementation to not rely on the second call to the Factory for `getPool()`. This upgrade could be an instance where it would save you unnecessary network calls using an endpoint, which if it were a paid endpoint, would mean money saved. 🙂

P.S. this is actually something we always challenge interns with, but don't mention it initially because there is still value in getting this far on your own.